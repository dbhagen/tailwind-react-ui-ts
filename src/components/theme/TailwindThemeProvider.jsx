import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
