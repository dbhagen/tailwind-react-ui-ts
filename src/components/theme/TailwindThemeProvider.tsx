import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
