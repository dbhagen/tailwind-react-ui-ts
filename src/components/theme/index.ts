export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
