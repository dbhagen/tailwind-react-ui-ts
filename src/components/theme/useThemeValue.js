import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
