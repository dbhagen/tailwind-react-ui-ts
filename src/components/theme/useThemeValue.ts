import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, value: string, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
