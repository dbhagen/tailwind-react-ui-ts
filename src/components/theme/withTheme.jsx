import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
