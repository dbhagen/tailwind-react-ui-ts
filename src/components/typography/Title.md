```jsx
<Title size={6}>Title 6</Title>
<Title size={5}>Title 5</Title>
<Title size={4}>Title 4</Title>
<Title size={3}>Title 3</Title>
<Title size={2}>Title 2</Title>
<Title size={1}>Title 1</Title>
```
