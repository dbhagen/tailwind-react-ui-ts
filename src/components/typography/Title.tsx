import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Text } from '../primitives'

const Title = ({
  theme,
  children,
  is,
  size,
  subtitle,
  flush,
  level,
  ...rest
}) => {
  const hLevel = level || Math.max(7 - size, 1)
  const element = is || `h${hLevel}`

  let ariaProps = {}

  if (!subtitle && element !== 'string' && !/h[1-6]/i.test(element)) {
    ariaProps = {
      role: 'heading',
      'aria-level': hLevel,
    }
  }

  return (
    <Text
      is={element}
      {...ariaProps}
      leading="tight"
      font={[
        theme.text.family[subtitle ? 'subtitle' : 'title'],
        subtitle ? 'medium' : 'bold',
      ]}
      text={[
        theme.text.size.title[size - 1],
        subtitle ? theme.textColors.body : theme.textColors.emphasis,
      ]}
      m={!flush ? { b: theme.spacing.md } : undefined}
      {...rest}
    >
      {children}
    </Text>
  )
}

Title.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  level: PropTypes.number,
}

Title.defaultProps = {
  children: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  level: undefined,
}

export { Title as component }
export default withTheme(Title)
