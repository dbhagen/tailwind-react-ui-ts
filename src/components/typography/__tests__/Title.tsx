import React from 'react'
import { shallow } from 'enzyme'

import { component as Title } from '../Title'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(<Title {...props}>Title</Title>)

  return {
    props,
    wrapper,
  }
}

describe('Title', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ size: 3 })

    expect(wrapper).toMatchSnapshot()
  })
})
