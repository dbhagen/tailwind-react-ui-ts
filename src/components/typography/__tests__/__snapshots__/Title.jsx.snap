// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title renders matching snapshot 1`] = `
<Text
  bold={false}
  font={
    Array [
      "sans",
      "bold",
    ]
  }
  is="h4"
  leading="tight"
  loose={false}
  m={
    Object {
      "b": 4,
    }
  }
  text={
    Array [
      "2xl",
      "black",
    ]
  }
  tight={false}
>
  Title
</Text>
`;
